### ISSUE TITLE
length of normal is not equal to 1

### ISSUE THREAD

**Issue by howardchina** [DurrLab/C3VD#5]
Hi!
I load the normal from "cecum_t1_a_under_review/c1v1/275_normals.tiff",
undistort the image using the single hole camera model,
and rescale XY to -1,1 and Z to 0,1.

And the normal is saved in a order of (H, W, C).

Then I calculate the length of normal like:
`np.sum(x * x, -1, keepdims=True)`

I don't know why the result is not 1 at all position.
Some of then are 0.4 and about 1% length are shorter than 1.

Second.
I have another stupid question. Why Z of normal ranges between 0 and 1 instead of -1 and 0?
Because the forward direction of Z is point from camera to surface in Fig.3. So why normal is not point from surface to your camera. And in this case, Z should range from -1 to 0 I guess?

**Comment by tbobrow1** [comment]
Hi @howardchina,

I just checked that particular image, and I do not see any normal values near 0.4. Can you share a reproducible script with the undistortion included so I can debug it? The interpolation used during undistortion will cause changes in the normal lengths, but I doubt this is why you are getting values equal to 0.4. The normals are rendered as 32 bits but saved with a bit depth of 16 bits, which will cause some minor errors. I usually rescale the normals to unitary after loading them so that all of the normals are exactly equal to length 1.0.

Regarding your question about the surface normal convention: I am using the standard "face-forward" convention, with the Z component scaled 0.0 to 1.0, with 1.0 pointing toward the camera (and therefore, opposite the direction of the camera's Z-axis).

**Comment by howardchina** [comment]
Hi, @tbobrow1 
I revisited my code and find I did't open it as 16 bits but 8 bits instead. So I revised my code, and followed your suggestion that I removed the undistortion. But finally the length of norm is still smaller than 1.0. So I paste my code and output below.

code:
```import numpy as np
import matplotlib.pyplot as plt
import cv2

root = "/home/data/datasets/c3vd/cecum_t1_a_under_review/c1v1/"
normal = root + "275_normals.tiff"

normal = cv2.imread(normal, -1)
bgr = normal.copy()
mask = bgr.sum(-1) > 0
x = bgr[:, :, 2] / (2**16-1) * 2 - 1
y = bgr[:, :, 1] / (2**16-1) * 2 - 1
z = bgr[:, :, 0] / (2**16-1)
n = np.stack([x, y, z], axis=-1)
len = np.sum(n * n, -1, keepdims=True) ** 0.5
print(len[mask])
print(len[mask].min(), len[mask].max())
plt.hist(len[mask])
plt.show()
```

output:
```
[[0.9658919 ]
 [0.9658919 ]
 [0.96343102]
 ...
 [0.9177058 ]
 [0.9177058 ]
 [0.9177058 ]]
0.6093190715978383 1.0000424000194452
```

![image](https://user-images.githubusercontent.com/6837044/231641850-1a49f23e-eb13-43d6-91dc-5039d61c08fd.png)

**Comment by tbobrow1** [comment]
Hi @howardchina,

Thank you for sending this code snippet and for allowing me to look into the normal length discrepancy. The code you shared is correct, and I am also getting many normals with a magnitude greater than/lesser than 1.0. I dug back through the raycasting code, and I have identified a small error in the transformation of the computed surface normals from world space to camera space. This is likely the cause of the erroneous values you are seeing, and it only affects the ground truth surface normal output, not the other ground truth components (depth, flow, etc). I am working on correcting the bug now, and I will verify this evening that this fixes the error you are seeing. If everything checks out, I will re-render the surface normals for all of the sequences tomorrow. I should be able to have them uploaded to the data repo by tomorrow evening.

Thank you for catching this and bringing it to my attention. I'll tag you here once I have this corrected.

**Comment by tbobrow1** [comment]
@howardchina I have corrected the error and re-rendered the ground truth surface normal frames. The z-components are now scaled -1 to 1 like the x- and y-components. I also switched the surface normal coordinate frame in screen space to match the camera coordinate system shown in Figure 3 of the paper to avoid confusion; +x now points to the right, +y points down, and +z points into the screen. Revised links/preview videos will go live on the webpage tonight once they finish uploading.

Thanks again for pointing this out. I will leave this issue open for a few days in case you identify any other problems once the files upload.
